include/net/xia_ether.h

```diff
@@ -0,0 +1,347 @@
+};
+
+static inline struct fib_xid_ether_local *fxid_lether(struct fib_xid *fxid)
+{
+	return likely(fxid)
+		? container_of(fxid, struct fib_xid_ether_local, xel_common)
+		: NULL;
+}
+
+/* Main ETHERs */
+struct ether_interface{
+	struct net_device			*dev;
+	atomic_t					refcnt;
+	int 						dead;
+	struct rcu_head				rcu_head;
+
+	atomic_t					neigh_cnt;
+	spinlock_t					interface_lock;
+	struct hh_cache				*chdr_head;
+	struct list_head			list_interface_common_addr;
+};
+
+struct interface_addr{
+	struct fib_xid_ether_main 	*mfxid;
+	struct list_head 			interface_common_addr;
+	struct net_device	 		*outgress_interface;
+	struct rcu_head				rcu_head;
+
+	u8 		ha[MAX_ADDR_LEN];
+};
+
+struct fib_xid_ether_main {
+	struct xip_dst_anchor	xem_anchor;
+	struct interface_addr 	*neigh_addr;
+	struct net_device	 	*host_interface;
+	int 					xem_dead;
+	struct hh_cache			cached_hdr;
+	rwlock_t				chdr_lock;
+	int						(*output)(struct fib_xid_ether_main *, struct sk_buff *);
+
+	/* WARNING: @xhm_common is of variable size, and
+	 * MUST be the last member of the struct.
+	 */
+	struct fib_xid		xem_common;
+};
+
+struct xia_header_ops {
+	int	(*create) (struct sk_buff *skb, struct net_device *dev,
+			   unsigned short type, const void *daddr,
+			   const void *saddr, unsigned int len);
+	int	(*parse)(const struct sk_buff *skb, unsigned char *haddr);
+	int	(*cache)(const struct fib_xid_ether_main *mfxid, struct hh_cache *hh, __be16 type);
+	void	(*cache_update)(struct hh_cache *hh,
+				const struct net_device *dev,
+				const unsigned char *haddr,
+			    const int type);
+};
+
+int xia_ether_header_cache(const struct fib_xid_ether_main *mfxid, struct hh_cache *hh, __be16 type)
+{
+	struct ethhdr *eth;
+	const struct net_device *dev = mfxid->host_interface;
+
+	eth = (struct ethhdr *)
+	    (((u8 *) hh->hh_data) + (HH_DATA_OFF(sizeof(*eth))));
+
+	if (type != htons(ETH_P_XIP))
+		return -1;
+
+	eth->h_proto = type;
+	memcpy(eth->h_source, dev->dev_addr, ETH_ALEN);
+	memcpy(eth->h_dest, mfxid->neigh_addr->ha, ETH_ALEN);
+	hh->hh_len = ETH_HLEN;
+	return 0;
+}
+
+void xia_ether_header_cache_update(struct hh_cache *hh,
+			     const struct net_device *dev,
+			     const unsigned char *haddr,
+			     const int type)
+{
+	struct ethhdr *eth;
+	eth = (struct ethhdr *)
+		    (((u8 *) hh->hh_data) + (HH_DATA_OFF(sizeof(*eth))));
+	if(type)
+		memcpy(eth->h_dest , haddr, ETH_ALEN);
+	else
+		memcpy(eth->h_source, dev->dev_addr, ETH_ALEN);
+}
+
+const struct xia_header_ops xia_ether_hdr_ops ____cacheline_aligned = {
+	.create		= eth_header,
+	.parse		= eth_header_parse,
+	.cache		= xia_ether_header_cache,
+	.cache_update	= xia_ether_header_cache_update,
+};
+
+static void mfxid_update_hhs(struct fib_xid_ether_main *mfxid, const int type)
+{
+	struct hh_cache *hh;
+	hh = &mfxid->cached_hdr;
+
+	if (hh->hh_len) {
+		write_seqlock_bh(&hh->hh_lock);
+		xia_ether_hdr_ops.cache_update(hh, mfxid->host_interface, mfxid->neigh_addr->ha, type);
```

Comment:
Break this line of code into multiple lines. The general guideline is that each line of code should be less than ``80`` characters.